### PR TITLE
Convert CMQ to QQ migration blog post to a guide

### PR DESCRIPTION
**Pull request by pstack2021**
…ntation to RabbitMQ OSS docs

### CODE REVIEW

**Comment by pstack2021** [comment]
@binarin can you please review the documentation, which is based on the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/) and the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#) do we need to include more information from this doc here such as the **Federate the New and Existing (now QQs) Queues** sections and more information from **Update Consumers to Use a Quorum Queue in the new Virtual Host** ??

To be honest, I think users will find this documentation hard to use. 

@MirahImage and @michaelklishin I have also included you as reviewers so please take a look if time permits. 

@binarin should Karl and Michal Kuratczyk be involved in this review? I notice the commented on the google doc.

This is the first draft of documentation, I can obviously edit it further based on your comments and @michaelklishin we can decide where to put it when reviews are complete (i.e. part of the existing QQ information or out on its own linking to QQ and CMQ information. 

Many thanks in advance
Paula

File: site/migrate-mcq-to-qq.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+* It is also important to note that migrated applications should be thoroughly tested against quorum queues because the behaviour can differ under the load and in edge cases. 
+
+## Deciding which Migration Route to take: Compatibility Considerations
+
+Incompatible features can be either referenced in policies or in the source code. RabbitMQ completes strict validation of arguments for queue declaration and consumption. Therefore, for migration, you must clean up all information about incompatible features in the source code. For some of these features, it is as simple as just removing corresponding strings, refer to [Mirrored Classic Queue Features that can be removed from Source Code or moved to a Policy](#mcq-features-to-remove), but other features require a change in the way queues are being used, refer to [Mirrored Classic Queue Features that require Changes in the Way the Queue is Used](#mcq-changes-way-queue-is-used).
```
Comment (binarin) [inline site/migrate-mcq-to-qq.md:37]: For my ear "completes validations" sounds unusual. I've only heard variants like"performs strict validation" or "strictly validates" before.

File: site/migrate-mcq-to-qq.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+```
+### `x-cancel-on-ha-failover` for Consumers
+
+Mirrored queues consumers can be [automatically
+cancelled](https://www.rabbitmq.com/ha.html#cancellation) when a queue
+leader fails over. This can cause loss of information about which messages were sent to which consumer, and result in the same messages being sent again (duplicate messages). 
+
+Quorum queues will not have the same results in this situation i.e. duplicate messages are not sent again except when there is a node failure, or messages are resent for inflight messages when the consumer is cancelled or the channel is closed.
+
+In summary, with mirrored classic queues, you can observe possible duplicates in some cases. With quorum queues, this observation is not possible but duplicates messages can still happen (less frequently) for some of the reasons they happened when using mirrored classic queues.
```
Comment (binarin) [inline site/migrate-mcq-to-qq.md:113]: I think that first sentence says that "observation" = "duplicates". Then the second on is contradictory, saying that "observation is not possible", "duplicates can still happen".

Some more emphasis is needed on the fact that application still needs to be prepared for duplicates, something along the lines:

With CMQs you can observe possible duplicates in some cases. Some of these cases were covered by `x-cancel-on-ha-failover`, and others weren't. Most of the cases covered by `x-cancel-on-ha-failover` went away with QQs. But the ones that aren't covered are still here, and your application needs to be aware / to be able to handle duplicates anyway.

File: site/migrate-mcq-to-qq.md (hunk 1 above)
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:37]: Thanks @binarin changed to "strictly validates arguments....

File: site/migrate-mcq-to-qq.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+-->
+
+## Migrating Mirrored Classic Queues to Quorum Queues
+Quorum Queues are now the optimum choice of queue type replacing mirrored classic queues. This information explains why, the reasons why you should migrate from mirrored classic queues to quorum queues, the ways to handle features during the migration, and includes procedures for some of the migration routes you can take. 
+
+You should migrate to mirrored classic queues for the following reasons:
+
+* Classic mirrored queues were deprecated in RabbitMQ version 3.9. They will be removed completely in RabbitMQ version 4.0.
+* Quorum queues can sustain much higher throughput levels in almost all use cases. A quorum queue can sustain a 30000 message throughput (using 1kb messages), while offering high levels of data safety, and replicating data to all 3 nodes in a cluster. Classic mirrored queues only offer a third of that throughput and provide much lower levels of data safety. 
+* Quorum queues are more reliable, faster for most workloads. and require little maintenance. For more information, go to [Quorum Queues](https://www.rabbitmq.com/quorum-queues.html).
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:25]: Fixed, thanks @binarin

File: site/migrate-mcq-to-qq.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+**Note**: 
+
+* While quorum queues are the best choice of queue type, they are not 100% compatible feature wise with mirrored classic queues. When you are deciding about whether to migrate from mirrored classic queues to quorum queues, it is recommended to review the [quorum queue documentation](https://www.rabbitmq.com/quorum-queues.html) first, you can review the [feature matrix table](https://www.rabbitmq.com/quorum-queues.html#feature-matrix) which provides a comparison of both queue types (mirrored classic queues beside quorum queues)
+
+* The level of complexity involved in migrating from mirrored classic queues to quorum queues depends on the features that are currently being used by the mirrored classic queues. Some of the steps require some configuration changes (refer to [Mirrored Classic Queue Features that require Changes in the Way the Queue is Used](#mcq-changes-way-queue-is-used)), while others require changes in the way an application interacts with RabbitMQ (refer to [Mirrored Classic Queue Features that can be removed from Source Code or moved to a Policy](#mcq-features-to-remove)).
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:31]: @binarin Is this review comment correct? They look correct to me? please advise - thanks

File: site/migrate-mcq-to-qq.md (hunk 2 above)
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:113]: Updates made for this comment, please review again @binarin

File: site/migrate-mcq-to-qq.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md]: @binarin Updated to make wording consistent between this section and the next one.

**Review by binarin — CHANGES_REQUESTED** [review]

File: site/migrate-mcq-to-qq.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+### Queue Length Limit overflow set to `reject-publish-dlx`
+
+The queue length exceeded with `overflow` set to [`reject-publish-dlx`](https://rabbitmq.com/maxlength.html#overflow-behaviour) is not supported by quorum queues. The `reject-publish-dlx` value is not supported.
+
+With mirrored classic queues, publishing to a full queue with `reject-publish-dlx` resulted in RabbitMQ republishing a rejected message to a dead letter exchange. With quorum queues, to apply the same logic, you must change `reject-publish-dlx` to `reject-publish`. Then, handle negative acknowledgements: after getting a negative acknowledgement, the application must publish the message again to a different exchange.
+
+To find out if  `overflow` set to `reject-publish-dlx` is configured for the mirrored classic queues you want to migrate, check for the `x-max-priority` string in the list of queues output that is provided by running the command in [Finding the Mirrored Classic Queues for Migration](#find-mcq) or you can also search for the ``reject-publish-dlx`` string in the source code.
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:88]: Thanks @binarin - DONE

**Comment by pstack2021** [comment]
Hi 

@kjnilsson and @mkuratczyk - This is the Migrating MCQ to QQ official documentation. I have created it from the the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/) and the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#). @binarin has reviewed it and I have implemented @binarin feedback (thanks very much @binarin for answering my slack queries as I created this documentation and your review). I know you have both commented on the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#) and the technical details in the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/) were also reviewed so please take a look if time permits.

@MirahImage and @edbyford 
Can you please also review this documentation, I do have concerns about the usability of it, will a customer be able to take it and migrate with it successfully using either of the migration routes provided in the doc (for example for the [Migrate in Place](https://github.com/rabbitmq/rabbitmq-website/blob/pstack-migrate-mcq-to-qq-docs/site/migrate-mcq-to-qq.md#migrate-in-place) route, we give them 6 steps, we don't provide any specific details/commands for those steps)? As far as I can remember @edbyford customers (think it was Dell) have seen the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#) in the past, I think you had a meeting with them some time back?
Perhaps an option here is to host this documentation on https://www.rabbitmq.com/documentation.html for now and host to VMware docs at a later time...see what feedback comes in on it? OR can we get some internal testing done on this migration process using this documentation to clarify it further and provide more detailed steps? 
**I would appreciate your input on this by the end of this week (Friday, the 2nd of June if possible), if you would like to me to host a meeting to discuss further, please let me know and I will organise that.**

Many thanks in advance
Paula.

**Review by binarin — APPROVED** [review]

File: site/migrate-mcq-to-qq.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+code, or move them to a policy.
+
+### Lazy Queues
+
+Mirrored classic queues can optionally operate in [lazy
+mode](https://rabbitmq.com/lazy-queues.html#overview). For quorum
+queues this is the only way of operation. The best way to handle the migration of these mirrored classic queues
+is to move `x-queue-mode` from the source code to a policy.
+
+### Transcient Queues
```
Comment (MirahImage) [inline site/migrate-mcq-to-qq.md:129]: Transient

File: site/migrate-mcq-to-qq.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+### Exclusive queues
+
+[Exclusive queues](https://www.rabbitmq.com/queues.html#exclusive-queues) are not mirrored. An attempt to create an exclusive mirrored queue creates a non-mirrored queue so the mirroring policy is ignored.
+
+For exclusive queues that are not mirrored, you must decide whether to leave the queue as an exclusive queue or change it to a replicated queue during migration. Therefore, you must be careful not to make exclusive queue declarations with an explicit `x-queue-type: quorum` argument.
+
+## General Prerequisites before Migrating to Quorum Queues
+
+1. A minimum of 3 nodes in the RabbitMQ cluster is required (there is no reason to use quorum queues with a smaller amount of replicas).
```
Comment (MirahImage) [inline site/migrate-mcq-to-qq.md:147]: This sentence is simply not true. While we recommend a minimum of 3 nodes for highly available quorum queues, it is perfectly reasonable to use quorum queues with a single node cluster, and there are some valid circumstances under which a user may decide to do so.

Comment (MirahImage) [reply site/migrate-mcq-to-qq.md:147]: I would recommend replacing this sentence with the following:
1. A RabbitMQ cluster with an odd number of nodes (for high availability, a minimum of 3 nodes is required).

File: site/migrate-mcq-to-qq.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+rabbitmqctl set_parameter shovel migrate-QUEUE_TO_MIGRATE \
+  '{"src-protocol": "amqp091", "src-uri": "amqp:///OLD_VHOST", "src-queue": "QUEUE_TO_MIGRATE",
+    "dest-protocol": "amqp091", "dest-uri": "amqp:///NEW_VHOST", "dest-queue": "QUEUE_TO_MIGRATE"}'
+```
+
+After the queue is drained, the shovel can be deleted:
+```bash
+rabbitmqctl clear_parameter shovel migrate-QUEUE_TO_MIGRATE
+```
+## <a id="migrate-in-place" class="anchor" href="#migrate-in-place">Migrate in Place</a>
```
Comment (MirahImage) [inline site/migrate-mcq-to-qq.md:269]: The steps for the in-place migration are very vague. Is it worth including them?

**Review by MirahImage — CHANGES_REQUESTED** [review]

File: site/migrate-mcq-to-qq.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+## Migrate your RabbitMQ Mirrored Classic Queues to Quorum Queues
+
+Quorum Queues are now the optimum choice of queue type replacing mirrored classic queues. This information explains why, the reasons why you should migrate from mirrored classic queues to quorum queues, the ways to handle features during the migration, and includes procedures for some of the migration routes you can take.
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:20]: Thanks @mkuratczyk, good point.

How about rewording to: "Which is better: mirrored classic queues or quorum queues? quorum queues are the better choice. This information explains why, the reasons why you should migrate from mirrored classic queues to quorum queues, the ways to handle features during the migration, and includes procedures for some of the migration routes you can take. "

I am now only making the comparison between MCQs and QQs....

File: site/migrate-mcq-to-qq.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+* Quorum queues can sustain much higher throughput levels in almost all use cases. A quorum queue can sustain a 30000 message throughput (using 1kb messages), while offering high levels of data safety, and replicating data to all 3 nodes in a cluster. Classic mirrored queues only offer a third of that throughput and provide much lower levels of data safety. 
+* Quorum queues are more reliable, faster for most workloads, and require little maintenance. For more information, go to [Quorum Queues](https://www.rabbitmq.com/quorum-queues.html). 
+
+**Note**: 
+
+* While quorum queues are the best choice of queue type, they are not 100% compatible feature wise with mirrored classic queues. When you are deciding about whether to migrate from mirrored classic queues to quorum queues, it is recommended to review the [quorum queue documentation](https://www.rabbitmq.com/quorum-queues.html) first, you can review the [feature matrix table](https://www.rabbitmq.com/quorum-queues.html#feature-matrix) which provides a comparison of both queue types (mirrored classic queues beside quorum queues)
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:30]: Thanks @mkuratczyk, I will reword to: "While quorum queues are a better queue type when compared to mirrored classic queues, they are......"

File: site/migrate-mcq-to-qq.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+### `x-cancel-on-ha-failover` for Consumers
+
+Classic mirrored queues consumers can be [automatically cancelled](https://www.rabbitmq.com/ha.html#cancellation) when a queue
+leader fails over. This can cause loss of information about which messages were sent to which consumer, and result in the same messages being sent again (duplicate messages). 
+
+Some of the cases for duplicate messages are covered by `x-cancel-on-ha-failover` and others are not. Most of the cases covered by `x-cancel-on-ha-failover` do not exist with quorum queues but those that are not covered are still there. Therefore, your application must be able to handle duplicates, which it should be able to do anyway.
+
+## <a id="mcq-features-to-remove" class="anchor" href="#mcq-features-to-remove">Mirrored Classic Queue Features that can be removed from Source Code or moved to a Policy</a>
+
+The following features don't complete any function when quorum queues are
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:118]: Thanks @mkuratczyk "are ignored" is best to use there, I will change.

File: site/migrate-mcq-to-qq.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkuratczyk) [reply site/migrate-mcq-to-qq.md:147]: We had a discussion about this. There are some cases where single node QQs make sense. Perhaps we should say something like "A minimum of 3 nodes in the RabbitMQ cluster is required for high availability (2-node quorum queues are a very bad idea and are not supported, to learn more: https://www.rabbitmq.com/quorum-queues.html#quorum-requirements)"

File: site/migrate-mcq-to-qq.md (hunk 8 above)
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:269]: Thanks @MirahImage who can give me these details including the commands to run to complete these steps ? Can someone test it and then provide these details to me?

Comment (MirahImage) [reply site/migrate-mcq-to-qq.md:269]: I think it's not possible for us to provide commands for these steps, because they are user-dependent

Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:269]: So @MirahImage when you say, what do you mean? 

> steps for the in-place migration are very vague. Is it worth including them

go into more detail for each of these:

Stop the consumers and producers.
Shovel the messages to a new temporary queue.
Delete the old queue.
Create a new quorum queue with the same name as the original queue.
Shovel the contents of the temporary queue to the new quorum queue.
Configure the consumers to use x-queue-type of quorum and they can be started.

?? who can provide more detail here? or should I root around RMQ OSS docs to find that info? 
Thanks

File: site/migrate-mcq-to-qq.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+The only option for transcient queues then will be exclusive queues. This only affects the durability of queue definitions. Messages can still be marked transient.
+
+You must make a decision about transcient queues before migration, is the content of the 
+queue important enough to get availability guarantees of quorum queues, or is it better to downgrade the transcient queue to a classic non-mirrored queue (classic mirrored queues are being removed but classic non-mirrored queues will still be available).
+
+### Exclusive queues
+
+[Exclusive queues](https://www.rabbitmq.com/queues.html#exclusive-queues) are not mirrored. An attempt to create an exclusive mirrored queue creates a non-mirrored queue so the mirroring policy is ignored.
+
+For exclusive queues that are not mirrored, you must decide whether to leave the queue as an exclusive queue or change it to a replicated queue during migration. Therefore, you must be careful not to make exclusive queue declarations with an explicit `x-queue-type: quorum` argument.
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:143]: @mkuratczyk Does this make more sense: 

You do not need to complete any migration tasks for exclusive queues. Exclusive queues are not mirrored even if the policy indicates that they are. Also, it is not possible to create an exclusive quorum queue. 

For exclusive queues, however, you must decide whether to leave the queue as exclusive or change it to a replicated queue during migration. Be careful not to make exclusive queue declarations with an explicit `x-queue-type: quorum` argument.

File: site/migrate-mcq-to-qq.md (hunk 7 above)
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:147]: Thanks @MirahImage I will update to this based on @mkuratczyk reply to the same information above: 

1. A RabbitMQ cluster with an odd number of nodes. A minimum of 3 nodes in the RabbitMQ cluster is required for high availability

Question though @MirahImage when you say "odd number" can that be any odd number up to what??? should we be specific here? Thanks

File: site/migrate-mcq-to-qq.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+Consumers of the migrated queues can now access the new queues by updating the connection parameters to connect to the new virtual host. The federation links start to pull in messages from the original queues.
+
+As with a blue-green cluster, after all consumers are migrated, you might need to also [add shovels to move the backlog of the original queues to the new queues](#shovel-remaining-messages) more efficiently than federation. For more information, refer to [Drain Messages](https://rabbitmq.com/blue-green-upgrade.html#drain-messages).
+
+### Update the Producers to Use the Exchanges in the new Virtual Host
+
+Once the original queues are empty (or nearly empty if you do not require full message ordering), the producers should be stopped and reconfigured to use the new queue declarations and virtual host like the consumers, and restarted. Federated exchanges in the old virtual host should also be stopped and equivalent exchanges should be added in the new virtual host. The original queues can be removed once they are empty and no messages are passing through them.
+
+Under sufficient system, load messages from the old virtual host will
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:250]: @mkuratczyk I pulled that sentence: "Under sufficient system load messages from the old virtual host will not be picked up. " from the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/). I will remove it, maybe it should be removed from the blog post also cc: @binarin ??

File: site/migrate-mcq-to-qq.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+used. The best way to handle these features is to remove them from the source
+code, or move them to a policy.
+
+### Lazy Queues
+
+Mirrored classic queues can optionally operate in [lazy
+mode](https://rabbitmq.com/lazy-queues.html#overview). For quorum
+queues this is the only way of operation. The best way to handle the migration of these mirrored classic queues
+is to move `x-queue-mode` from the source code to a policy.
+
```
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:128]: Thanks @mkuratczyk updating the entire section to the following as per `[slack conversation.](https://vmware.slack.com/archives/D03B3J54TN1/p1685624840212189)` 

### Lazy Queues
Quorum queues do not support lazy mode (x-queue-mode=lazy).

To migrate mirrored lazy classic queues, remove the `x-queue-mode=lazy` declaration argument or remove it from the policy if it is set via a policy.

[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:128]: Thanks @mkuratczyk updating the entire section to the following as per [slack conversation.](https://vmware.slack.com/archives/D03B3J54TN1/p1685624840212189)

### Lazy Queues
Quorum queues do not support lazy mode (x-queue-mode=lazy).

To migrate mirrored lazy classic queues, remove the `x-queue-mode=lazy` declaration argument or remove it from the policy if it is set via a policy.

File: site/migrate-mcq-to-qq.md (hunk 8 above)
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:269]: ok for now @MirahImage we will leave the steps as is but perhaps we should consider documenting the steps for 1 use case scenario - is there the more popular one that we know of????

File: site/migrate-mcq-to-qq.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md:250]: Thanks @mkuratczyk updated to: "Under sufficient system load, messages from the old virtual host will
not be picked up....."

**Comment by pstack2021** [comment]
> Hi
> 
> @kjnilsson and @mkuratczyk - This is the Migrating MCQ to QQ official documentation. I have created it from the the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/) and the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#). @binarin has reviewed it and I have implemented @binarin feedback (thanks very much @binarin for answering my slack queries as I created this documentation and your review). I know you have both commented on the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#) and the technical details in the [blog post](https://blog.rabbitmq.com/posts/2023/02/quorum-queues-migration/) were also reviewed so please take a look if time permits.
> 
> @MirahImage and @edbyford Can you please also review this documentation, I do have concerns about the usability of it, will a customer be able to take it and migrate with it successfully using either of the migration routes provided in the doc (for example for the [Migrate in Place](https://github.com/rabbitmq/rabbitmq-website/blob/pstack-migrate-mcq-to-qq-docs/site/migrate-mcq-to-qq.md#migrate-in-place) route, we give them 6 steps, we don't provide any specific details/commands for those steps)? As far as I can remember @edbyford customers (think it was Dell) have seen the [google doc](https://docs.google.com/document/d/1IW8PHRCqkSkXQjsuawGC-8rp4_d-M_XdTNGf1JkFaAs/edit#) in the past, I think you had a meeting with them some time back? Perhaps an option here is to host this documentation on https://www.rabbitmq.com/documentation.html for now and host to VMware docs at a later time...see what feedback comes in on it? OR can we get some internal testing done on this migration process using this documentation to clarify it further and provide more detailed steps? **I would appreciate your input on this by the end of this week (Friday, the 2nd of June if possible), if you would like to me to host a meeting to discuss further, please let me know and I will organise that.**
> 
> Many thanks in advance Paula.

**UPDATE**
Thanks @MirahImage and @mkuratczyk for reviewing, updates for your comments are in https://github.com/rabbitmq/rabbitmq-website/commit/6bee4cc1264dae10264256b7e5a8caa717509c2e

@kjnilsson I know you are on vacation this week. Feel free to review the latest documentation next week when you return.

@edbyford, perhaps you can take a quick look when you return and provide some thoughts based on my original comments/concerns above.

Thanks again all.

File: site/migrate-mcq-to-qq.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+[Exclusive queues](https://www.rabbitmq.com/queues.html#exclusive-queues) are not mirrored. An attempt to create an exclusive mirrored queue creates a non-mirrored queue so the mirroring policy is ignored.
+
+For exclusive queues that are not mirrored, you must decide whether to leave the queue as an exclusive queue or change it to a replicated queue during migration. Therefore, you must be careful not to make exclusive queue declarations with an explicit `x-queue-type: quorum` argument.
+
+## General Prerequisites before Migrating to Quorum Queues
+
+1. A minimum of 3 nodes in the RabbitMQ cluster is required (there is no reason to use quorum queues with a smaller amount of replicas).
+2. The Management plugin should be running on at least one node. It is used to export/import definitions for a single host,
+   which simplifies definitions cleanup. (`rabbitmqadmin` CLI command is also using the plugin behind the scenes).
+3. The [Shovel plugin](https://www.rabbitmq.com/shovel.html) should be enabled.
```
[the remark this answers is not in the document]
Comment (michaelklishin) [reply site/migrate-mcq-to-qq.md]: "The Shovel plugin can be used to move the backlog of original messages to the new queues. Shovels can be created programmatically via an HTTP API extension or via the management UI."

[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md]: Thank you @michaelklishin

[the remark this answers is not in the document]
Comment (pstack2021) [reply site/migrate-mcq-to-qq.md]: FYI @michaelklishin now updated to: 

The [Shovel plugin](https://www.rabbitmq.com/shovel.html) can be used to move the backlog of original messages to the new queues. Shovels can be created programmatically using a HTTP API extension or using the RabbitMQ Management UI.

**Comment by pstack2021** [comment]
@michaelklishin @mkuratczyk @MirahImage can you please approve this PR if you are happy with the latest updates I have made for your feedback, many thanks.

**Review by mkuratczyk — APPROVED** [review]

**Comment by pstack2021** [comment]
@MirahImage and @michaelklishin if you can approve this PR if happy with the current doc that would be great, many thanks.

**Review by MirahImage — APPROVED** [review]
